Stop response download from signal handler (#4559)
scrapy/scrapy: docs/topics/exceptions.rst

Here's a list of all exceptions included in Scrapy and their usage.

CancelRequest
-------------

.. exception:: CancelRequest

> Raised to indicate that no further bytes should be downloaded for a request.

Reviewer: I think the approach of raising an exception from a singnal handler is good, but we need to explain better the semantics, and check how it works with other parts of Scrapy.

1. Signal name. For cancelled request I'd not expect callback to be called. Here we're not cancelling a request (request is already sent); we're stopping/aborting download of the response. 
2. Scrapy can also abort download on different conditions, such as max response size. I think in this case errback is called, not callback. Should we mention this in docs? There is also DOWNLOAD_FAIL_ON_DATALOSS option which look related, though I'm not sure it should affect the behavior of the cancel exception. 
3. I'm not sure what's better, call a callback or an errback, and if there is a single best approach. Maybe there should be 2 exceptions (or a bool flag in exception), one which make download successful, and another one for errors. E.g. in the FAQ entry you're writing "detect early that its response won't be useful", but if response is not useful, it shouldn't go to a callback.
4. Are there other places where this exception can be raised? I don't have anything in mind, but the CancelRequest exception look similar to IgnoreRequest which can be raised by a downloader middlewares; this can be confusing. Are there use cases for catching CancelRequest from other signals?

Author: Those are very good points, thank you.

(1) I don't particularly love the current name, and I agree with the fact that the "request is already sent", so it's not completely accurate. What about `(Abort/Interrupt/Stop)Response`?

(2/3) I think that since this behaviour would be controlled explicitly by user code, and not because of an unexpected error, we should call the assigned callback. That said, providing a way to fire the errback instead sounds fine to me too. What about an `errback` boolean argument in the exception, which defaults to `False`?

(4) I don't think we could catch this exception anywhere else, since the signal is directly sent from the download handler. Responses in middlewares are already complete.

Reviewer 2: (1) Based on the documentation, `(Abort/Cancel/Interrupt/Stop)Download`?

Author: Sounds even better. At point point I thought about `(Verb)(Response/Body)Download`, since what we interrupt is technically only the body, but that was too verbose, and probably unnecessary.

Author: I went for `StopDownload`. Also reworded the docs from "the response might not be useful" to "the already downloaded content might be enough". Given that (and my comment below about this being "controlled explicitly by user code") I think it'd be good to fire the callback instead of the errback. Thoughts?

Reviewer 2: Adding a `DOWNLOAD_FAIL_ON_DOWNLOAD_STOP` (or `DOWNLOAD_FAIL_ON_STOP`?) setting may make sense, like `DOWNLOAD_FAIL_ON_DATALOSS`.

I have no strong on what the default behavior should be. While the error is caused by users, as @elacuesta said, I can imagine users desiring both behaviors:
- If the signal is used to stop downloads once the desired information has been downloaded, `callback` makes sense.
- If the signal is used to stop downloads as soon as they are detected as bad responses based on content, `errback` makes sense to cause a retry (as it happens for data loss with `DOWNLOAD_FAIL_ON_DATALOSS=True`).

This exception is raised to indicate an unsupported feature.

StopDownload
-------------

.. versionadded:: 2.2

> .. exception:: StopDownload(fail=False)

Reviewer: What do you think about either the removing the default value, or making it True by default?

Argument for removing the default: `StopDownload(fail=False)` is easier to understand, and not much longer to write.

Argument for making fail=True by default: this feature can be seen as a much more flexible DOWNLOAD_MAXSIZE, and DOWNLOAD_MAXSIZE causes errback to be called.

> .. exception:: StopDownload(errback=False)

Reviewer: Sorry, one more query, about naming :) `errback` can be confused with errback argument of Request; previously all "errback" arguments were functions, not bools. I wonder if it is possible to have a flag name without this issue. Some ideas (more ideas are welcome):

* `StopDownload(fail=True)`, `StopDownload(fail=False)`
* `StopDownload(ok=True)`, `StopDownload(ok=False)`
* `StopDownload` and `CancelDownload`
* `StopDownload` and `IgnoreRequest`

Author: I like `StopDownload(fail)`. Thanks!

Raised from a :class:`~scrapy.signals.bytes_received` signal handler to
indicate that no further bytes should be downloaded for a response.

> The ``fail`` boolean parameter indicates whether or not the resulting partial
> response should be handled by the request errback. In this case, the response
> will be available in the ``response`` attribute of the received
> :class:`~twisted.python.failure.Failure` object.
> ``fail`` is ``False`` by default, which means the response will be handled
> by the request callback. Note that ``fail`` is a keyword-only argument.
> 
> Regardless of which method handles the response, it will contain
> ``"download_stopped"`` in its :attr:`Response.flags` attribute.

Reviewer: ```suggestion
The ``fail`` boolean parameter indicates whether or not the resulting partial
response should be handled by the request callback (fail=False, default) or by
the request errback (fail=True).

* If ``StopDownload(fail=False)`` is raised, request callback is called. Response
  which is passed to the callback could have its body truncated: the body
  contains only bytes received before the StopDownload exception is raised.
* When ``StopDownload(fail=True)`` exception is raised, request errback is called. 
  In this case response (possibly with a partial body) is available as the 
  ``response`` attribute of the received :class:`~twisted.python.failure.Failure` 
  object.

Note that ``fail`` is a keyword-only argument.

Regardless of which method handles the response, the response object has
``"download_stopped"`` in its :attr:`Response.flags` attribute.
```

Reviewer: While writing the suggestion above, I realized we don't specify which data would response.body actually contain. In particular, would it include the current data chunk received in bytes_received signal?

Author: Updated, thanks!

This exception is raised to indicate an unsupported feature.

StopDownload
-------------

.. versionadded:: 2.2

.. exception:: StopDownload(fail=True)

Raised from a :class:`~scrapy.signals.bytes_received` signal handler to
indicate that no further bytes should be downloaded for a response.

The ``fail`` boolean parameter controls which method will handle the resulting
response:

* If ``fail=True``, the request errback is called. The response object is
  available as the ``response`` attribute of the received
  :class:`~twisted.python.failure.Failure` object. This is the default behaviour.

* If ``fail=False``, the request callback is called instead.

Note that ``fail`` is a keyword-only parameter.

> In both cases, the response could have its body truncated: the body contains

Reviewer: What do you think about moving a note about kwarg-only argument somewhere else, to not separate 2 cases from "In both cases, ..."?

all bytes received up until the exception is raised, including the bytes

> received in the signal handler that raises the exception. The response object
> willl have ``"download_stopped"`` in its :attr:`Response.flags` attribute.

Reviewer: ```suggestion
received in the signal handler that raises the exception. Also, the response 
object is marked with ``"download_stopped"`` in its :attr:`Response.flags` 
attribute.
```

This exception is raised to indicate an unsupported feature.

StopDownload
-------------

.. versionadded:: 2.2

.. exception:: StopDownload(fail=True)

Raised from a :class:`~scrapy.signals.bytes_received` signal handler to
indicate that no further bytes should be downloaded for a response.

The ``fail`` boolean parameter controls which method will handle the resulting
response:

* If ``fail=True``, the request errback is called. The response object is
  available as the ``response`` attribute of the received
  :class:`~twisted.python.failure.Failure` object. This is the default behaviour.

* If ``fail=False``, the request callback is called instead.

In both cases, the response could have its body truncated: the body contains
all bytes received up until the exception is raised, including the bytes

> received in the signal handler that raises the exception. Also, the response

Reviewer: How hard would it be to add a test for it?

Author: Done in fbe915d9cbd8ab801a679b05265cff416764d558
